tests/e2e/contracts.go

```diff
@@ -89,9 +94,159 @@ func ContractsTest(sc *RuntimeScenario, log *logging.Logger, conn *grpc.ClientCo
+				Symbol:   "OAS20TEST",
+				Decimals: 2,
+				InitialBalances: []oas20.InitialBalance{
+					{
+						Address: types.NewAddress(signer.Public()),
+						Amount:  *quantity.NewFromUint64(10_000),
+					},
+				},
+			},
+		},
+		[]types.BaseUnits{},
+	).
+		SetFeeGas(1_000_000).
+		AppendAuthSignature(signer.Public(), nonce+4)
+	_ = tb.AppendSign(ctx, signer)
+	var instanceOas20 contracts.InstantiateResult
+	if err := tb.SubmitTx(ctx, &instanceOas20); err != nil {
+		return fmt.Errorf("failed to instantiate OAS20 contract: %w", err)
+	}
+
+	// Tansfer some tokens.
+	tb = ct.Call(
+		instanceOas20.ID,
+		&oas20.Request{
+			Transfer: &oas20.Transfer{
+				To:     testing.Charlie.Address,
+				Amount: *quantity.NewFromUint64(10),
+			},
+		},
+		[]types.BaseUnits{},
+	).
+		SetFeeGas(1_000_000).
+		AppendAuthSignature(signer.Public(), nonce+5)
+	_ = tb.AppendSign(ctx, signer)
+	if err := tb.SubmitTx(ctx, &rawResult); err != nil {
+		return fmt.Errorf("failed to call OAS20 transfer: %w", err)
+	}
+
+	// TODO: watch emitted event.
+
+	var response *oas20.Response
+	if err := ct.Custom(
+		ctx,
+		client.RoundLatest,
+		instanceOas20.ID,
+		&oas20.Request{
+			Balance: &oas20.Balance{
+				Address: testing.Charlie.Address,
+			},
+		},
+		&response,
+	); err != nil {
+		return fmt.Errorf("failed to query OAS20 balance: %w", err)
+	}
+	if response.Balance == nil {
+		return fmt.Errorf("invalid OAS20 query balance response: %v", response)
+	}
+	if response.Balance.Balance.Cmp(quantity.NewFromUint64(10)) != 0 {
+		return fmt.Errorf("unexpected OAS20 query balance response: %v", response.Balance)
+	}
+
+	// Send some tokens to hello contract.
+	tb = ct.Call(
+		instanceOas20.ID,
+		&oas20.Request{
+			Send: &oas20.Send{
+				To:     instance.ID,
+				Amount: *quantity.NewFromUint64(10),
+				Data:   13, // Specifies by how much the counter should be incremented.
+			},
+		},
+		[]types.BaseUnits{},
+	).
+		SetFeeGas(1_000_000).
+		AppendAuthSignature(signer.Public(), nonce+6)
+	_ = tb.AppendSign(ctx, signer)
+	if err := tb.SubmitTx(ctx, &rawResult); err != nil {
+		return fmt.Errorf("failed to call OAS20 send: %w", err)
+	}
+	// Sending to hello contract bumps the counter.
+	counter += 13
+
+	// Check counter.
+	tb = ct.Call(
+		instance.ID,
+		map[string]map[string]string{
+			"say_hello": {
+				"who": "e2e test",
+			},
+		},
+		[]types.BaseUnits{},
+	).
+		SetFeeGas(1_000_000).
+		AppendAuthSignature(signer.Public(), nonce+7)
+	_ = tb.AppendSign(ctx, signer)
+	if err := tb.SubmitTx(ctx, &rawResult); err != nil {
+		return fmt.Errorf("failed to call hello contract: %w", err)
+	}
+
+	if err := cbor.Unmarshal(rawResult, &result); err != nil {
+		return fmt.Errorf("failed to decode contract result: %w", err)
+	}
+
+	if result["hello"]["greeting"] != fmt.Sprintf("hello e2e test (%d)", counter) {
 		return fmt.Errorf("unexpected result from contract: %+v", result)
 	}
+	// Calling say_hello bumps the counter.
+	// counter++
```

Comment:
Is this commented out because this would otherwise throw the ineffective lint error?

Reply:
yeah, i guess i could also use the nolint directive